pkg/derivation/output_path.go

```diff
@@ -0,0 +1,90 @@
+package derivation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+)
+
+// OutputPaths returns a map of output names to the calculated store paths.
+// For more than trivial derivations, calculating these requires some
+// substitutions of other referred InputDerivations, so a store needs to be passed,
+// that can be used to look up other Derivations.
+func (d *Derivation) OutputPaths(store Store) (map[string]string, error) {
+	derivationName, ok := d.Env["name"]
+	if !ok {
+		// asserted by Validate
+		panic("env 'name' not found")
+	}
+
+	// we only want to call partial drv hash if we're not a fixed output
+	var partialDrvHash string
+
+	if fixed := d.GetFixedOutput(); fixed == nil {
+		var buf bytes.Buffer
+
+		err := d.writeDerivation(&buf, true, store)
+		if err != nil {
+			return nil, err
+		}
+
+		h := sha256.New()
+
+		_, err = h.Write(buf.Bytes())
+		if err != nil {
+			return nil, err
+		}
+
+		partialDrvHash = hex.EncodeToString(h.Sum(nil))
+	}
+
+	// We populate a new map of outputPaths, which is what we return in the end of the function
+	outputPaths := make(map[string]string)
+
+	var err error
+	for outputName, o := range d.Outputs {
+		outputPaths[outputName], err = o.outputPath(outputName, derivationName, partialDrvHash)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return outputPaths, nil
+}
+
+// getSubstitutionHash is producing a hex-encoded hash of the current derivation.
+// It needs a store, as it does some substitution on the way.
+func (d *Derivation) getSubstitutionHash(store Store) (string, error) {
```

Comment:
I think we want to move this method to the Derivation store.

Right now we recursively calculate this potentially multiple times.

This only needs to be calculated once for each derivation, so stores might populate some internal cache, either while calling `Put()`, or lazily on-first-access.

Reply:
I really don't like this to be a function on the drv store implementation.
I think we need to reconsider this.

Reply:
The Derivation store function only is a trampoline to `func (d *Derivation) OutputPaths(store Store)`.

All hash computation logic is still in `Derivation`, we just pipe the requests through the Derivation Store interface, so the calculation can be cached, and we can store the intermediate results to recuse less.

Reply:
-> https://github.com/nix-community/go-nix/pull/58#discussion_r891231379